model/riscv_pmp_regs.sail

```diff
@@ -154,15 +163,31 @@ function pmpReadCfgReg(n) = {
 }
 
 /* Helpers to handle locked entries */
-function pmpLocked(cfg: Pmpcfg_ent) -> bool =
-   cfg.L() == 0b1
-
-function pmpTORLocked(cfg: Pmpcfg_ent) -> bool =
-   (cfg.L() == 0b1) & (pmpAddrMatchType_of_bits(cfg.A()) == TOR)
-
-function pmpWriteCfg(cfg: Pmpcfg_ent, v: bits(8)) -> Pmpcfg_ent =
-  if pmpLocked(cfg) then cfg
-  else Mk_Pmpcfg_ent(v & 0x9f)  // Bits 5 and 6 are zero.
+function pmpLockBit(cfg: Pmpcfg_ent) -> bool = cfg.L() == 0b1
+   
+function pmpLocked(cfg: Pmpcfg_ent) -> bool = (cfg.L() == 0b1) & (mseccfg.RLB() == 0b0)
+
+function pmpTORLocked(cfg: Pmpcfg_ent) -> bool = (cfg.L() == 0b1) & (mseccfg.RLB() == 0b0) & (pmpAddrMatchType_of_bits(cfg.A()) == TOR)
+  
+function pmpWriteCfg(cfg: Pmpcfg_ent, v: bits(8)) -> Pmpcfg_ent = {
+  let RWX_001 : Pmpcfg_ent = Mk_Pmpcfg_ent(v[7 .. 3] @ 0b1 @ 0b0 @ 0b0);
+  let RWX_010 : Pmpcfg_ent = Mk_Pmpcfg_ent(v[7 .. 3] @ 0b0 @ 0b1 @ 0b0);
+  let RWX_011 : Pmpcfg_ent = Mk_Pmpcfg_ent(v[7 .. 3] @ 0b1 @ 0b1 @ 0b0);
+  let RWX_101 : Pmpcfg_ent = Mk_Pmpcfg_ent(v[7 .. 3] @ 0b1 @ 0b0 @ 0b1);
+  let val_to_write : Pmpcfg_ent = Mk_Pmpcfg_ent(v & 0xff);
+  if pmpLocked(cfg) then cfg                                          /* If locked then configuration is unchanged */
+  else {
+    if ((mseccfg.MML() == 0b1) & (mseccfg.RLB() == 0b0) & (v[7] == bitone)) then {
+      if ((val_to_write == RWX_001) | (val_to_write == RWX_010) | (val_to_write == RWX_011) | (val_to_write == RWX_101))
+      then cfg                                                        /* If condition meets configuration remains same which means pmpcfg writes are ignored */
+      else Mk_Pmpcfg_ent(v & 0x9f)                                    /* Bits 5 and 6 are zero */
+    } else {
+    if (mseccfg.MML() == 0b0) & (v[0] == bitzero) & (v[1] == bitone)
+    then Mk_Pmpcfg_ent(v & 0x9c)                                      /* Bits 5 and 6 are zero and Bits 0 and 1 are zero because of RW=01 reserved combination */
+    else Mk_Pmpcfg_ent(v & 0x9f)                                      /* Bits 5 and 6 are zero */
+    }
+  }
```

Comment:
This is not what I said to do. Construct _one_ Pmpcfg_ent and use field accessors in your if statements. No bitvector indexing, and no. As for updating, use field updates, don't then go back to the bitvector, the whole point is to make the code self-documenting in terms of the fields it's touching.

Reply:
**Fixed** ... The only concern your now be, line 173 of _riscv_pmp_regs.sail_ .But for constructing legal Pmpcfg_ent  vector/bit concatenation wasn't working which is why I did, like that otherwise everything is according to your comments. For reference, you can look into that snippet
![saill](https://user-images.githubusercontent.com/98050888/216312769-eca21686-e01d-4023-8180-a8556b8ee682.png)